com.unity.multiplayer.mlapi/Runtime/Core/NetworkObject.cs

```diff
@@ -504,6 +509,196 @@ internal void InvokeBehaviourOnGainedOwnership()
             }
         }
 
+        private bool m_IsReparented;
+        private ulong? m_LatestParent;
+        private Transform m_CachedParent;
+
+        internal void InvokeBehaviourOnNetworkObjectParentChanged(NetworkObject parentNetworkObject)
+        {
+            for (int i = 0; i < ChildNetworkBehaviours.Count; i++)
+            {
+                ChildNetworkBehaviours[i].OnNetworkObjectParentChanged(parentNetworkObject);
+            }
+        }
+
+        internal void SetCachedParent(Transform parentTransform)
+        {
+            m_CachedParent = parentTransform;
+        }
+
+        private void OnTransformParentChanged()
+        {
+            if (transform.parent == m_CachedParent)
+            {
+                return;
+            }
+
+            if (NetworkManager == null || !NetworkManager.IsListening)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new NotListeningException($"{nameof(NetworkManager)} is not listening, start a server or host before reparenting"));
+                return;
+            }
+
+            if (!NetworkManager.IsServer)
+            {
+                transform.parent = m_CachedParent;
+                Debug.LogException(new NotServerException($"Only the server can reparent {nameof(NetworkObject)}s"));
```

Comment:
how would this work with a client authoritative NetworkTransform?

Reply:
`NetworkTransform` is a `NetworkBehaviour` and a component — similarly, `NetworkBehaviour` conceptually a component to `NetworkObject`. So, in this context, we're talking about reparenting a `NetworkObject`, not a `NetworkTransform` or anything.
`NetworkTransform` could still replicate its state (pos,rot,sca etc.) however it wants. If it requires a parental change, it has to ask server, more specifically `NetworkObject` to change its parent.

Reply:
mmm so the NetworkTransform has to wait a full RTT for the server to confirm the parent change? 
if my character is jumping on the back of a moving predicted car, I wouldn't be able to use parenting client side until RTT ms? I'd have a full RTT where my character would jitter around on the car until parenting is activated? that's not good.

Reply:
you could have 2 `NetworkObject`s and enable one, disable the other and keep going without any issues and potentially even without any parenting during the transition.

Reply:
So if I have 100 players and 100 cars, each car would need 100 disabled players as children on them in case the player jumps on them? And I'd need to copy my player state to each of these cars?
That doesn't seem very users first :(

Reply:
the first question, does MLAPI can handle 100 concurrent players with 100 cars?

Reply:
second question, why do you keep "player state" on that particular `NetworkObject` and nowhere else?

Reply:
third question, would coupling features together to build a feature and limiting some other potential cases is more user-first than this and how can we agree on which one is more user-first here?

Reply:
1. can you say it'll never do?
2. as user, it'd be pretty simple. I have my player, I move it around, it contains my own state. This way I have a self contained player.
3. I'm not saying we should couple NetworkTransform and NetworkObject at all. I'm just talking about allowing NetworkObject to change parenting client side without waiting for a full RTT, in a client authoritative model.

Reply:
1. I can say, it's not in our scope or in any of our acceptance criteria etc.
2. you're explaining a specific scenario we may or may not support (which could be a good user-need to start with)
3. there is no authority model in MLAPI other than Ownership, except the ad-hoc `NetworkTransform` internal one — and I'm hesitant to expand the scope of this feature/delivery.

Reply:
MLAPI has supported client authority for a while though. It's not formalized like we're aiming to do it, but it's still supported. Dropping that support is kind of a big deal.
I'm guessing since you're arguing against it there's a good technical reason for not supporting client authority in that reparenting case?
To be noted, I'm not talking about shared authority, more about owner authority, where only an owner could change the parent, but that change could come either from the client or from the server. I understand with shared authority it can become a mess very fast, but would it be that complex to allow clients changing parenting too?

Reply:
> MLAPI has supported client authority for a while though.

where specifically?

> Dropping that support is kind of a big deal.

checkout the ownership code, look into `NetworkObject::ChangeOwnership(ulong newOwnerClientId)` API.
following your definition, MLAPI already dropped client authority long before then.

> I'm guessing since you're arguing against it there's a good technical reason for not supporting client authority in that reparenting case?

I'm not arguing against anything here. I'm double-checking your assumptions and understandings.

> To be noted, I'm not talking about shared authority, more about owner authority, where only an owner could change the parent, but that change could come either from the client or from the server. I understand with shared authority it can become a mess very fast, but would it be that complex to allow clients changing parenting too?

ownership model (`IsOwner`, `OwnerClientId`, `ChangeOwnership()` etc.) does not play a role here. the code user writes on the server-side can decide to change with regards to the object's owner or regardless of it, we do not assume or try to guess. however, `ChangeOwnership()` API works only on the server-side because the server is the ultimate source of truth. we are following a very similar concept here: which `NetworkObject` is parent/child of which `NetworkObject` is determined, controlled and stored by the server — depending on what user writes, it may or may not be affected by ownership or any other arbitrary conditions but the execution must happen on the server-side as it is for ownership.

Reply:
So since AutoObjectParentSync is a per-NetworkObject property, could one enable the kind of client-authoritative parenting @SamuelBellomo is describing by controlling this flag on the desired objects?

Reply:
you can opt-out from `AutoObjectParentSync`, yes. however, client-driven/authoritative object parenting sounds like a completely different feature to me (and I'd even argue if that's needed given that we're a client-server model).